pkg/backend/endpoint_resolver.go

```diff
@@ -0,0 +1,186 @@
+				for _, epAddr := range epSubset.NotReadyAddresses {
+					if epAddr.TargetRef == nil || epAddr.TargetRef.Kind != "Pod" {
+						continue
+					}
+					epPod, err := r.findPodByReference(ctx, svc.Namespace, *epAddr.TargetRef)
+					if err != nil {
+						return nil, err
+					}
+					if isPodMeetCriteria(epPod, resolveOpts.UnreadyPodInclusionCriteria) {
+						endpoints = append(endpoints, buildPodEndpoint(epPod, epAddr, epPort))
+					}
+				}
+			}
+		}
+	}
+
+	return endpoints, nil
+}
+
+func (r *defaultEndpointResolver) ResolveNodePortEndpoints(ctx context.Context, svcKey types.NamespacedName, port intstr.IntOrString, opts ...EndpointResolveOption) ([]NodePortEndpoint, error) {
+	resolveOpts := EndpointResolveOptions{
+		NodeSelector: labels.Nothing(),
+	}
+	resolveOpts.ApplyOptions(opts)
+	svc, svcPort, err := r.findServiceAndServicePort(ctx, svcKey, port)
+	if err != nil {
+		return nil, err
+	}
+
+	if svc.Spec.Type != corev1.ServiceTypeNodePort && svc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return nil, errors.Errorf("service type must be either 'NodePort' or 'LoadBalancer': %v", svcKey)
+	}
+	svcNodePort := svcPort.NodePort
+	nodeList := &corev1.NodeList{}
+	if err := r.k8sClient.List(ctx, nodeList, client.MatchingLabelsSelector{Selector: resolveOpts.NodeSelector}); err != nil {
+		return nil, err
+	}
+
+	var endpoints []NodePortEndpoint
+	for i := range nodeList.Items {
+		node := &nodeList.Items[i]
+		if !k8s.IsNodeReady(node) {
+			continue
+		}
+		instanceID, err := extractNodeInstanceID(node)
+		if err != nil {
+			return nil, err
+		}
+		endpoints = append(endpoints, buildNodePortEndpoint(node, instanceID, svcNodePort))
+	}
+
+	return endpoints, nil
+}
+
+func (r *defaultEndpointResolver) findServiceAndServicePort(ctx context.Context, svcKey types.NamespacedName, port intstr.IntOrString) (*corev1.Service, corev1.ServicePort, error) {
+	svc := &corev1.Service{}
+	if err := r.k8sClient.Get(ctx, svcKey, svc); err != nil {
+		return nil, corev1.ServicePort{}, err
+	}
+	svcPort, err := k8s.LookupServicePort(svc, port)
+	if err != nil {
+		return nil, corev1.ServicePort{}, err
+	}
+
+	return svc, svcPort, nil
+}
+
+func (r *defaultEndpointResolver) findPodByReference(ctx context.Context, namespace string, podRef corev1.ObjectReference) (*corev1.Pod, error) {
+	pod := &corev1.Pod{}
+	podKey := types.NamespacedName{Namespace: namespace, Name: podRef.Name}
+	if err := r.k8sClient.Get(ctx, podKey, pod); err != nil {
+		return nil, err
+	}
+	return pod, nil
+}
+
+func isPodMeetCriteria(pod *corev1.Pod, criteria []PodPredicate) bool {
+	for _, criterion := range criteria {
+		if !criterion(pod) {
+			return false
+		}
+	}
+	return true
+}
+
+func extractNodeInstanceID(node *corev1.Node) (string, error) {
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		return "", errors.Errorf("providerID is not specified for node: %s", node.Name)
+	}
+
+	p := strings.Split(providerID, "/")
+	return p[len(p)-1], nil
```

Comment:
what if providerID doesn't contain '/' character?

Reply:
if it don't contain '/', then full providerID is returned :D